File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/model/network/NetworkIntentService.java
package com.dbottillo.mtgsearchfree.model.network;

import android.app.IntentService;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.content.LocalBroadcastManager;

import com.dbottillo.mtgsearchfree.R;
import com.dbottillo.mtgsearchfree.model.TCGPrice;
import com.dbottillo.mtgsearchfree.util.LOG;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserFactory;

import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;

public class NetworkIntentService extends IntentService {
    public static final String EXTRA_PARAMS = "NetworkIntentService.EXTRA_PARAMS";
    public static final String EXTRA_ID = "NetworkIntentService.EXTRA_ID";
    public static final String EXTRA_CARD_NAME = "NetworkIntentService.EXTRA_CARD_NAME";
    public static final String EXTRA_SET_NAME = "NetworkIntentService.EXTRA_SET_NAME";
    public static final String REST_RESULT = "com.dbottillo.mtgsearch.network.REST_RESULT";
    public static final String REST_URL = "com.dbottillo.mtgsearch.network.REST_URL";

    public NetworkIntentService() {
        super("NetworkIntentService");
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        TCGPrice res = null;
        Bundle extras = intent.getExtras();
        Bundle params = extras.getParcelable(EXTRA_PARAMS);
        String cardName = "";
        String setName = "";
        String idRequest = "";

        if (params != null) {
            cardName = params.getString(EXTRA_CARD_NAME, "");
            setName = params.getString(EXTRA_SET_NAME, "").replace(" ", "%20");
            idRequest = params.getString(EXTRA_ID);
            cardName = cardName.replace(" ", "%20").replace("Æ", "ae");
        }

        String url = "http://partner.tcgplayer.com/x3/phl.asmx/p?pk=MTGCARDSINFO&s=" + setName + "&p=" + cardName;

        LOG.INSTANCE.d("loading price for card " + cardName);
        try {
            res = doNetworkRequest(url);
        } catch (Exception e) {
            LOG.INSTANCE.e(e);
        }

        if (res != null && (res.getLowprice() == null || res.getLowprice().equalsIgnoreCase("0")) && setName != null && setName.length() > 0) {
            url = "http://partner.tcgplayer.com/x3/phl.asmx/p?pk=MTGCARDSINFO&s=&p=" + cardName;
            LOG.INSTANCE.d("try again without set for card " + cardName);
            try {
                res = doNetworkRequest(url);
            } catch (Exception e) {
                LOG.INSTANCE.e(e);
            }
        }

        if (res == null) {
            res = new TCGPrice();
            res.setError(getApplicationContext().getString(R.string.price_error));
        }

        Intent intentRes = new Intent(idRequest);
        intentRes.putExtra(REST_RESULT, res);
        intentRes.putExtra(REST_URL, url);
        LocalBroadcastManager.getInstance(this).sendBroadcast(intentRes);
    }

    private TCGPrice doNetworkRequest(String url) throws Exception {
        TCGPrice tcgPrice = new TCGPrice();
        URL uri = new URL(url);
        HttpURLConnection urlConnection = (HttpURLConnection) uri.openConnection();
        urlConnection.setRequestMethod("GET");
        urlConnection.setRequestProperty("Connection", "close");

        if (urlConnection.getResponseCode() == 200) {
            InputStream in = urlConnection.getInputStream();

            XmlPullParserFactory xmlFactoryObject = XmlPullParserFactory.newInstance();
            XmlPullParser myparser = xmlFactoryObject.newPullParser();
            myparser.setInput(in, null);
            int event = myparser.getEventType();
            boolean isHiPrice = false;
            boolean isLowPrice = false;
            boolean isAvgPrice = false;
            boolean isLink = false;
            while (event != XmlPullParser.END_DOCUMENT) {
                String name = myparser.getName();
                if (event == XmlPullParser.START_TAG) {
                    isHiPrice = false;
                    isLowPrice = false;
                    isAvgPrice = false;
                    isLink = false;
                    if (name.equals("hiprice")) {
                        isHiPrice = true;
                    }
                    if (name.equals("avgprice")) {
                        isAvgPrice = true;
                    }
                    if (name.equals("lowprice")) {
                        isLowPrice = true;
                    }
                    if (name.equals("link")) {
                        isLink = true;
                    }

                } else if (event == XmlPullParser.TEXT) {
                    if (isHiPrice) {
                        tcgPrice.setHiPrice(myparser.getText());
                    }
                    if (isLowPrice) {
                        tcgPrice.setLowprice(myparser.getText());
                    }
                    if (isAvgPrice) {
                        tcgPrice.setAvgPrice(myparser.getText());
                    }
                    if (isLink) {
                        tcgPrice.setLink(myparser.getText());
                    }

                } else if (event == XmlPullParser.END_TAG) {
                    isHiPrice = false;
                    isLowPrice = false;
                    isAvgPrice = false;
                    isLink = false;

                }
                event = myparser.next();
            }
            if (tcgPrice.getHiPrice() == null) {
                tcgPrice.setError(getApplicationContext().getString(R.string.price_error));
            }
            in.close();
        } else {
            if (urlConnection.getResponseCode() == 500) {
                tcgPrice.setNotFound(true);
            }
            tcgPrice.setError(getApplicationContext().getString(R.string.price_error));
        }
        urlConnection.disconnect();
        return tcgPrice;
    }
}

File: MTGSearch/src/main/kotlin/com/dbottillo/mtgsearchfree/model/network/NetworkIntentService.kt
package com.dbottillo.mtgsearchfree.model.network

import android.app.IntentService
import android.content.Intent
import android.os.Bundle
import androidx.localbroadcastmanager.content.LocalBroadcastManager

import com.dbottillo.mtgsearchfree.R
import com.dbottillo.mtgsearchfree.model.TCGPrice
import com.dbottillo.mtgsearchfree.util.LOG

import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserFactory

import java.net.HttpURLConnection
import java.net.URL

class NetworkIntentService : IntentService("NetworkIntentService") {

    override fun onHandleIntent(intent: Intent) {
        var res: TCGPrice? = null
        val extras = intent.extras
        val params = extras?.getParcelable<Bundle>(EXTRA_PARAMS)
        var cardName = ""
        var setName: String? = ""
        var idRequest: String? = ""

        if (params != null) {
            cardName = params.getString(EXTRA_CARD_NAME, "")
            setName = params.getString(EXTRA_SET_NAME, "").replace(" ", "%20")
            idRequest = params.getString(EXTRA_ID)
            cardName = cardName.replace(" ", "%20").replace("Æ", "ae")
        }

        var url = "https://partner.tcgplayer.com/x3/phl.asmx/p?pk=MTGCARDSINFO&s=$setName&p=$cardName"

        LOG.d("loading price for card $cardName")
        try {
            res = doNetworkRequest(url)
        } catch (e: Exception) {
            LOG.e(e)
        }

        if (res != null && (res.lowprice == null || res.lowprice!!.equals("0", ignoreCase = true)) && setName != null && setName.isNotEmpty()) {
            url = "https://partner.tcgplayer.com/x3/phl.asmx/p?pk=MTGCARDSINFO&s=&p=$cardName"
            LOG.d("try again without set for card $cardName")
            try {
                res = doNetworkRequest(url)
            } catch (e: Exception) {
                LOG.e(e)
            }
        }

        if (res == null) {
            res = TCGPrice()
            res.setError(applicationContext.getString(R.string.price_error))
        }

        val intentRes = Intent(idRequest)
        intentRes.putExtra(REST_RESULT, res)
        intentRes.putExtra(REST_URL, url)
        LocalBroadcastManager.getInstance(this).sendBroadcast(intentRes)
    }

    @Throws(Exception::class)
    private fun doNetworkRequest(url: String): TCGPrice {
        val tcgPrice = TCGPrice()
        val uri = URL(url)
        val urlConnection = uri.openConnection() as HttpURLConnection
        urlConnection.requestMethod = "GET"
        urlConnection.setRequestProperty("Connection", "close")

        if (urlConnection.responseCode == 200) {
            val inputStream = urlConnection.inputStream

            val xmlFactoryObject = XmlPullParserFactory.newInstance()
            val myparser = xmlFactoryObject.newPullParser()
            myparser.setInput(inputStream, null)
            var event = myparser.eventType
            var isHiPrice = false
            var isLowPrice = false
            var isAvgPrice = false
            var isLink = false
            while (event != XmlPullParser.END_DOCUMENT) {
                val name = myparser.name
                if (event == XmlPullParser.START_TAG) {
                    isHiPrice = false
                    isLowPrice = false
                    isAvgPrice = false
                    isLink = false
                    if (name == "hiprice") {
                        isHiPrice = true
                    }
                    if (name == "avgprice") {
                        isAvgPrice = true
                    }
                    if (name == "lowprice") {
                        isLowPrice = true
                    }
                    if (name == "link") {
                        isLink = true
                    }
                } else if (event == XmlPullParser.TEXT) {
                    if (isHiPrice) {
                        tcgPrice.hiPrice = myparser.text
                    }
                    if (isLowPrice) {
                        tcgPrice.lowprice = myparser.text
                    }
                    if (isAvgPrice) {
                        tcgPrice.avgPrice = myparser.text
                    }
                    if (isLink) {
                        tcgPrice.link = myparser.text
                    }
                } else if (event == XmlPullParser.END_TAG) {
                    isHiPrice = false
                    isLowPrice = false
                    isAvgPrice = false
                    isLink = false
                }
                event = myparser.next()
            }
            if (tcgPrice.hiPrice == null) {
                tcgPrice.setError(applicationContext.getString(R.string.price_error))
            }
            inputStream.close()
        } else {
            if (urlConnection.responseCode == 500) {
                tcgPrice.isNotFound = true
            }
            tcgPrice.setError(applicationContext.getString(R.string.price_error))
        }
        urlConnection.disconnect()
        return tcgPrice
    }

    companion object {
        const val EXTRA_PARAMS = "NetworkIntentService.EXTRA_PARAMS"
        const val EXTRA_ID = "NetworkIntentService.EXTRA_ID"
        const val EXTRA_CARD_NAME = "NetworkIntentService.EXTRA_CARD_NAME"
        const val EXTRA_SET_NAME = "NetworkIntentService.EXTRA_SET_NAME"
        const val REST_RESULT = "com.dbottillo.mtgsearch.network.REST_RESULT"
        const val REST_URL = "com.dbottillo.mtgsearch.network.REST_URL"
    }
}
